Networking/Networking/ApplicationPassword/RequestAuthenticator.swift

```diff
@@ -0,0 +1,90 @@
+enum RequestAuthenticatorError: Error {
+    case applicationPasswordUseCaseNotAvailable
+    case applicationPasswordNotAvailable
+}
+
+/// Authenticates request
+///
+public struct RequestAuthenticator {
+    /// Credentials.
+    ///
+    let credentials: Credentials?
+
+    /// The use case to handle authentication with application passwords.
+    ///
+    private let applicationPasswordUseCase: ApplicationPasswordUseCase?
+
+    /// Sets up the authenticator with optional credentials and application password use case.
+    /// `applicationPasswordUseCase` can be injected for unit tests.
+    ///
+    init(credentials: Credentials?, applicationPasswordUseCase: ApplicationPasswordUseCase? = nil) {
+        self.credentials = credentials
+        let useCase: ApplicationPasswordUseCase? = {
+            if let applicationPasswordUseCase {
+                return applicationPasswordUseCase
+            } else if case let .wporg(username, password, siteAddress) = credentials {
+                return try? DefaultApplicationPasswordUseCase(username: username,
+                                                              password: password,
+                                                              siteAddress: siteAddress)
+            } else {
+                return nil
+            }
+        }()
+        self.applicationPasswordUseCase = useCase
+    }
+
+    func authenticate(_ urlRequest: URLRequest) throws -> URLRequest {
+        if isRestAPIRequest(urlRequest) {
+            return try authenticateUsingApplicationPasswordIfPossible(urlRequest)
+        } else {
+            return try authenticateUsingWPCOMTokenIfPossible(urlRequest)
+        }
+    }
+
+    func generateApplicationPassword() async throws {
+        guard let applicationPasswordUseCase = applicationPasswordUseCase else {
```

Comment:
nit: might look nicer
```suggestion
        guard let applicationPasswordUseCase else {
```

Reply:
Done in a3e21d8a6e7422a2fc1489974d61d0a8c7d71ab7